models/git_diff.go

```diff
@@ -56,12 +63,93 @@ type DiffSection struct {
 	Lines []*DiffLine
 }
 
+func diffToHtml(diffRecord []diffmatchpatch.Diff, lineType DiffLineType) template.HTML {
+	result := ""
+	for _, s := range diffRecord {
+		if s.Type == diffmatchpatch.DiffInsert && lineType == DIFF_LINE_ADD {
+			result = result + "<span class=\"added-code\">"+html.EscapeString(s.Text)+"</span>"
+		} else if s.Type == diffmatchpatch.DiffDelete && lineType == DIFF_LINE_DEL {
+			result = result + "<span class=\"removed-code\">"+html.EscapeString(s.Text)+"</span>"
+		} else if s.Type == diffmatchpatch.DiffEqual {
+			result = result + html.EscapeString(s.Text)
+		}
+	}
+	return template.HTML(result)
+}
+
+func (diffSection *DiffSection) GetLeftLine(idx int, sliceIdx int) *DiffLine {
```

Comment:
Sure.

To [compute the diff](https://github.com/andreynering/gogs/blob/bf11ad19eab0de5aef297a21ed58c6cd811a137c/models/git_diff.go#L140) I must have two line strings, the deleted and the added one. `GetLeftLine` try to get the deleted one when the actual is an add. `GetRightLine` is the oposite. [See](https://github.com/andreynering/gogs/blob/bf11ad19eab0de5aef297a21ed58c6cd811a137c/models/git_diff.go#L122-L137).

Unfortunaly, this isn't working when the left index is different of the right index. [Example](https://github.com/andreynering/gogs/commit/81ed5c4bee112beacd8e0f34b336b4f3b6c6a6c1#diff-478b0554d47378204e26103fd20ec06bL33): left is 33 and right 34 for the same line.

I tried to fix that, but it's tricky. Any help would be appreciated.